[doc][core][cgraph] Complete Compiled Graph docs (#51206)
ray-project/ray: doc/source/ray-core/compiled-graph/quickstart.rst

``asyncio`` support
-------------------

If your Compiled Graph driver is running in an ``asyncio`` event loop, use the ``async`` APIs to ensure that executing

> the Compiled Graph and getting the results doesn't block the event loop. This requires a few changes for now.

Reviewer: ```suggestion
the Compiled Graph and getting the results doesn't block the event loop.
```
Already stated that it requires using the async API's. Also the `for now` implies this is something that will change in the future, if it will, specify that it might?

First, pass ``enable_async=True`` to the ``dag.experimental_compile()``:

.. literalinclude:: ../doc_code/cgraph_quickstart.py
    :language: python
    :start-after: __cgraph_async_compile_start__
    :end-before: __cgraph_async_compile_end__

> Next, use `execute_async` to invoke the Compiled Graph. Calling ``await`` on ``execute_async`` will return once
> the input has been submitted, and it returns a future that can be used to get the result. Finally, 
> use `await` to get the result of the Compiled Graph.

Reviewer: ```suggestion
Next, use `execute_async` to invoke the Compiled Graph.
Then, call ``await`` on ``execute_async`` to wait until the input is submitted and to get a future.
Finally, call `await` on that future to get the result of the Compiled Graph.
```

Author: Thanks. I think your phrasing is more consistent within the context, while the original description (coming from alpha developer guide) emphasizes the effect of``await`` on ```execute_async``, so may be clearer to the user what they would expect. So I would keep it. Let me know if you feel otherwise and we can update.

:func:`ray.get() <ray.get>` also has a timeout parameter to set timeout on a per-call basis.

CPU to GPU communication
------------------------

> With classic Ray Core, passing ``torch.Tensors`` between actors can easily become expensive, especially

Reviewer: ```suggestion
With Ray Core, passing ``torch.Tensors`` between actors can become expensive, especially
```

Author: Going to keep "classic", Compiled Graph is still part of Ray Core, so it's clearer to distinguish

when transferring between devices. This is because Ray Core doesn't know the final destination device.
Therefore, you may see unnecessary copies across devices other than the source and destination devices.

Ray Compiled Graph ships with native support for passing ``torch.Tensors`` between actors executing on different

> devices. Developers can now use type hint annotations in the Compiled Graph declaration to indicate the final
> destination device of a ``torch.Tensor``. For example, let’s say we have an actor running on a GPU, and we
> want to pass a ``torch.Tensor`` to it.

Reviewer: ```suggestion
devices. You can use type hint annotations in the Compiled Graph declaration to indicate the final
destination device of a ``torch.Tensor``.
```
can also do something saying In Ray Core you'd do this like the following code sample. But this `For example` sentence feels out of place. Can also say a possible use case of this would be ...

Author: removed

.. literalinclude:: ../doc_code/cgraph_nccl.py
    :language: python
    :start-after: __cgraph_cpu_to_gpu_actor_start__
    :end-before: __cgraph_cpu_to_gpu_actor_end__

In Ray Core, if you try to pass a CPU tensor from the driver,
the GPU actor receives a CPU tensor:

.. testcode::

    # This will fail because the driver passes a CPU copy of the tensor, 
    # and the GPU actor also receives a CPU copy.
    ray.get(actor.process.remote(torch.zeros(10)))

With Ray Compiled Graph, you can annotate DAG nodes with type hints to indicate that there may be a ``torch.Tensor``
contained in the value:

.. literalinclude:: ../doc_code/cgraph_nccl.py
    :language: python
    :start-after: __cgraph_cpu_to_gpu_start__
    :end-before: __cgraph_cpu_to_gpu_end__

Under the hood, the Ray Compiled Graph backend copies the ``torch.Tensor`` to the GPU assigned to the ``GPUActor`` by Ray Core.

Of course, you can also do this yourself, but there are advantages to using Compiled Graph instead:

> - Ray Compiled Graph can minimize the number of data copies made. For example, passing from one CPU to

Reviewer: ```suggestion
- Ray Compiled Graph minimizes the number of data copies made. For example, passing from one CPU to
```

  multiple GPUs requires one copy to a shared memory buffer, and then one host-to-device copy per
  destination GPU.

- In the future, this can be further optimized through techniques such as memory pinning, using

>   zero-copy deserialization when the CPU is the destination, etc.

Reviewer: will the compiled graph be doing this optimization under the hood, is the idea that the user will get these improvements without code changes? If that's the point, make that more obvious

Author: I think the details are TBD. It may be fully automatic or requires minor code changes. This level of detail looks OK for now.
